### ISSUE TITLE
Support remote instances

### ISSUE THREAD

**Issue by shaaati** [Patryk27/lxd-snapper#8]
If I understand it correctly, lxd-snapper currently supports only local instances.
I currently have a backup process set up where my backup server interacts with another lxd remote instance. It manually creates snapshots and pulls them afterwards with `lxc copy --refresh`. I'd love to replace the manual snapshot process with lxd-snapper and I'd rather not move this bit to the productive server but instead keep it in one place.

Would it be possible to implement a config parameter that allows specifying a remote that a specific policy is applied to?
I imagine that there are many use cases for this, e.g., if people manage multiple different lxd instances.

**Comment by Patryk27** [comment]
Hi,

So if I understand correctly, with your current approach the backup server and the remote(s) both keep the snapshots -- the backup server acts as a coordinator doing something like `lxc snapshot remote:container` + `lxc copy --refresh remote:container container`, right?

**Comment by shaaati** [comment]
Exactly. I found that creating and copying snapshots is the easiest way to regularly conduct backups of lxd containers. I **think** (never tested this so far) that if I delete a snapshot on the remote, it should get deleted on the backup host upon the next run of `lxc copy --refresh` and therefore I keep all snapshots on both hosts.

In theory, this would allow for a central backup server that aggregates snapshots of multiple different remotes.

All I would need from lxd-snapper is a way to run it against a remote instead of locally. Since the same is possible with the manual `lxc snapshot` commands, I imagine this shouldn't be too hard to implement.

**Comment by Patryk27** [comment]
Yeah, this should be doable - I'll try preparing some proof of concept in a few days and let you know 🙂

**Comment by shaaati** [comment]
Exactly. I found that creating and copying snapshots is the easiest way to regularly conduct backups of lxd containers. I **think** (never tested this so far) that if I delete a snapshot on the remote, it should get deleted on the backup host upon the next run of `lxc copy --refresh` and therefore I keep all snapshots on both hosts.

In theory, this would allow for a central backup server that aggregates snapshots of multiple different remotes.

All I would need from lxd-snapper is a way to run it against a remote instead of locally. Since the same is possible with the manual `lxc snapshot` commands, I imagine this shouldn't be too hard to implement.

23.08.2022 19:25:56 Patryk Wychowaniec ***@***.***>:

**Comment by Patryk27** [comment]
(I remember about this feature, just haven't had as much time on my hands as I had imagined 😇 -- but hopefully things should clear out in a few days ™️ )

**Comment by shaaati** [comment]
Where on earth does my last comment (with the same content as before in August) come from? I most certainly did not reply to this thread by email. 🤔

Did you maybe reply to a mail and this got mixed up somehow?

Anyway, I appreciate your sign of life :). Take your time, it's not **that** time-critical for my. I am more than happy that you considered implementing this at all.

**Comment by Patryk27** [comment]
As for the notifications, I think GitHub's having some issues 😅 -- similar thing happened e.g. [on the nixpkgs repository](https://github.com/NixOS/nixpkgs/pull/168300#issuecomment-1274398313), where a new message appeared just yesterday, while the commenter said they wrote it months ago.

**Comment by aivanise** [comment]
Hi,

I've been following this with interest as I'm doing the same thing using syncoid and find it a bit cumbersome, so I'm looking to replace it with something else.

I've tried and on my servers if I keep doing lxc copy --refresh container backup: the snapshots from local will be synced with the backup all the time no matter what I do on the backup, so I'm not sure what is the point here, as I don't need to create any snapshots on the backup, I can even remove them all from the backup server and the copy --refresh will still sync them all on the next run

**Comment by shaaati** [comment]
> Hi,
> 
> I've been following this with interest as I'm doing the same thing using syncoid and find it a bit cumbersome, so I'm looking to replace it with something else.
> 
> I've tried and on my servers if I keep doing lxc copy --refresh container backup: the snapshots from local will be synced with the backup all the time no matter what I do on the backup, so I'm not sure what is the point here, as I don't need to create any snapshots on the backup, I can even remove them all from the backup server and the copy --refresh will still sync them all on the next run
> 

The idea is not to create further snapshots **on the backup host**, but to use the backup host as an orchestrator that creates snapshots **on the main server** and then pulls the new state over (minimizing the data that needs to be transferred, as only the difference since the last snapshots must be copied to the backup host).

Currently, I have a custom script that runs on the backup host. It creates a new snapshot for each container every day and then does an incremental sync. I recently ran into a situation where the relatively high amount of snapshots triggered a bug in LXD and therefore needed to come up with a way to reliably delete older snapshots so that the overall amount stays within reasonable numbers. That's when I found lxd-snapper, because why should I come up with an own retention mechanism for snapshots if there is nice software already doing the hard work? :)
In its current state, lxd-snapper cannot be used as a drop-in replacement in my script.

@aivanise: I hope this description of my (intended) workflow clarifies what I try to achieve with this issue.

**Comment by Patryk27** [comment]
Hi, @shaaati - I've just released lxd-snapper 1.3.0 that supports LXD remotes 🙂 

The only thing that I haven't implemented is automatic `lxc copy --refresh` (mostly because it's got a few switches and I wasn't sure which ones to use for the feature to be both maintainable and useful) -- but, fortunately, you can get that pretty easily by using hooks, so something like this should get you started:

``` yaml
hooks:
  on-instance-backed-up: 'lxc copy --refresh {{ remoteName }}:{{ instanceName }} {{ instanceName }}'
  on-instance-pruned: 'lxc copy --refresh {{ remoteName }}:{{ instanceName }} {{ instanceName }}'

remotes:
  - server-A
  - server-B
  - server-C

policies:
  all-servers:
    keep-last: 10
```

If you imagined this feature somewhat differently, don't hesitate to lemme know - I don't have use case for remotes personally, so I'd just guessed how this should work 🙂
